### ISSUE TITLE
new rule: no-import-reassign

### ISSUE THREAD

**Issue by xjamundx** [eslint/eslint#2948]
I'd like a rule that made it so that `import`ed variables could not be re-assigned or clobbered.

Recently I did something like this and eslint didn't complain:

```js
import x from "x";
function x() {}
x();
```

The problem is that the `import` way at the top was so far away from the `function` down below that I didn't notice it and reused the same name.

Are there any rules that address this already?

These are similar but do not cover my use case:
- http://eslint.org/docs/rules/no-native-reassign.html
- http://eslint.org/docs/rules/no-param-reassign.html
- http://eslint.org/docs/rules/no-shadow

Thanks and let me know if you can help :)

**Comment by btmills** [comment]
@nzakas is there anything in the spec that would prevent a situation like this in module scope, similar to how `let` and `const` redeclarations throw? Either way, I would think no-redeclare should catch this.

**Comment by xjamundx** [comment]
Okay yeah looks like https://github.com/eslint/eslint/blob/master/lib/rules/no-redeclare.js doesn't look for `ImportDeclaration` and maybe that's the issue?

**Comment by mysticatea** [comment]
Please see: #2903
The `no-redeclare` rule had not been checking module scopes, and it's fixed now.

But, I also think the `no-import-reassign` rule is needed :)

**Comment by xjamundx** [comment]
FYI babel does show an error with this code: `Line 2: Duplicate declaration "x"`.

It may be an espree issue if that's the case. Will check a few more sources.

**Comment by mysticatea** [comment]
On `master` branch:

.eslintrc

```json
{
    "ecmaFeatures": {"modules": true},
    "rules": {"no-redeclare": 2}
}
```

----

```
> git log -1
commit dffc8e01d20f263a0b8cb0edbe0e917e27a81b83
...
> echo "import x from 'x'; function x() {}" | node .\bin\eslint.js --stdin --reset

<text>
  1:30  error  x is already defined  no-redeclare

✖ 1 problem (1 error, 0 warnings)
```

**Comment by xjamundx** [comment]
Yup this is fixed in master. Will look into if it's possibly a parser issue as well. Thanks!

**Comment by vhiroki** [comment]
@mysticatea, it still doesn't complain of reassignments with `=` :
```
import x from "./some-module";
x = 123;
```

Is that right?
